### PR TITLE
[BEAM-755] Fix typo in DirectRunner dependenciesToScan

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @tgroh AND @lukecwik

### CODE REVIEW

**Comment by kennknowles** [comment]
This also needs a test jar dep. Hold for update.

**Comment by kennknowles** [comment]
OK good for review.

**Comment by kennknowles** [comment]
Test failures are internal failures of the Dataflow service, not this PR, as you might surmise by the lack of relation between them.

**Comment by kennknowles** [comment]
Thanks @tgroh. To @lcwik for second LGTM from a committer.

File: runners/direct-java/pom.xml [hunk 1]
```diff
@@ -83,7 +83,7 @@
               <failIfNoTests>true</failIfNoTests>
               <dependenciesToScan>
                 <dependency>org.apache.beam:beam-sdks-java-core</dependency>
-                <dependency>org.apache.beam:beam-runners-java-core</dependency>
```
Comment (lukecwik) [inline runners/direct-java/pom.xml]: Instead of adding it to the test scope, we should limit this to adding it as plugin specific dependencies if possible.
https://maven.apache.org/guides/mini/guide-configuring-plugins.html#Using_the_dependencies_Tag

This way it doesn't pollute the test scope of the regular project. We should follow this pattern for our other "integration" test setups.

Comment (kennknowles) [inline runners/direct-java/pom.xml]: A bit of scope creep, but it's nice, so I've done it.

File: runners/direct-java/pom.xml [hunk 2]
```diff
@@ -70,6 +70,14 @@
             <beamUseDummyRunner>false</beamUseDummyRunner>
           </systemPropertyVariables>
         </configuration>
+        <dependencies>
+          <dependency>
+            <groupId>org.apache.beam</groupId>
+            <artifactId>beam-runners-core-java</artifactId>
+            <type>test-jar</type>
+            <version>${project.version}</version>
```
Comment (kennknowles) [inline runners/direct-java/pom.xml]: If this version is left out, maven crashes on it, not taking the version from `dependencyManagement`. Yet in `runners-parent` it did not require a version. I don't know why.

File: runners/direct-java/pom.xml [hunk 3]
```diff
@@ -310,6 +318,13 @@
 
     <!-- test dependencies -->
     <dependency>
+      <groupId>org.apache.beam</groupId>
```
Comment (kennknowles) [inline runners/direct-java/pom.xml]: This is an actual dependency to pull in `WindowMatchers`.

Note that it is using `classifier` instead of `type`. I suspect that `classifiers` is almost always better than `test-jar` since it is more abstract, but that is more Maven work than I'd like to do right now.

File: runners/spark/pom.xml [hunk 4]
```diff
@@ -72,7 +72,7 @@
                     <parallel>none</parallel>
                     <failIfNoTests>true</failIfNoTests>
                     <dependenciesToScan>
-                      <dependency>org.apache.beam:java-sdk-all</dependency>
+                      <dependency>org.apache.beam:beam-sdks-java-core</dependency>
```
Comment (kennknowles) [inline runners/spark/pom.xml]: CC: @amitsela just tweaking this typo while I'm futzing with the Maven config. It might cause a trivial conflict elsewhere.

Comment (amitsela) [inline runners/spark/pom.xml]: LGTM, this is already fixed in #1055 but if that should go in now, I'll rebase and resolve my PR later.

**Comment by kennknowles** [comment]
Jenkins failure is an internal service error. I'll squash the fixups to kick it.

**Comment by kennknowles** [comment]
Please take another look when ready. Note that the three commits should not be squashed.

File: runners/pom.xml [hunk 5]
```diff
@@ -55,6 +55,13 @@
             <plugin>
               <groupId>org.apache.maven.plugins</groupId>
               <artifactId>maven-surefire-plugin</artifactId>
+              <dependencies>
+                <dependency>
+                  <groupId>org.apache.beam</groupId>
+                  <artifactId>beam-sdks-java-core</artifactId>
+                  <type>test-jar</type>
+                </dependency>
```
Comment (lukecwik) [inline runners/pom.xml]: This is only working because every runner depends on beam-sdks-java-core already, best to put the version in explicitly

File: runners/pom.xml [hunk 6]
```diff
@@ -55,6 +55,13 @@
             <plugin>
               <groupId>org.apache.maven.plugins</groupId>
               <artifactId>maven-surefire-plugin</artifactId>
+              <dependencies>
+                <dependency>
+                  <groupId>org.apache.beam</groupId>
+                  <artifactId>beam-sdks-java-core</artifactId>
+                  <type>test-jar</type>
```
Comment (lukecwik) [inline runners/pom.xml]: Shouldn't we depend on the shaded version of the test jar and not the vanilla one since beam-sdks-java-core does shading?

File: runners/direct-java/pom.xml [hunk 7]
```diff
@@ -83,7 +91,7 @@
               <failIfNoTests>true</failIfNoTests>
               <dependenciesToScan>
                 <dependency>org.apache.beam:beam-sdks-java-core</dependency>
-                <dependency>org.apache.beam:beam-runners-java-core</dependency>
+                <dependency>org.apache.beam:beam-runners-core-java</dependency>
```
Comment (lukecwik) [inline runners/direct-java/pom.xml]: Doesn't seem to work, I made sure that there was a NeedsRunner annotation on a test and it never executed.

Comment (kennknowles) [inline runners/direct-java/pom.xml]: Hmm, I'll take a look in a bit. It might be that the reorganized deps don't work well with `dependenciesToScan`.

File: runners/pom.xml (hunk 6 above)
Comment (kennknowles) [inline runners/pom.xml]: What is the diff to the pom for this? I guess I never even tried just changing this to a `classifier`. Is that known not to work?

Comment (lukecwik) [inline runners/pom.xml]: The main artifact in beam-sdks-java-core that is produced uses the shade plugin. For any dependency that the test uses directly that is not a dependency of this module and also repackaged, bundled and stripped in beam-sdks-java-core would cause the test code to fail with class not found exceptions.

Comment (kennknowles) [inline runners/pom.xml]: I see. Yes, tests of APIs that are not public need to be adjusted alongside the classes under test. I would actually expect that ROS tests don't do this.

But indeed, the main test-jar should match the main output jar. Users of runnable-on-service tests shouldn't have to reason about how the jar is produced. I'm sure we can make that work.

**Comment by kennknowles** [comment]
At this point, I don't think the approach of moving the deps around works quite right. I'll close this and open one that just fixes the typo and deps.
